### ISSUE TITLE
Fully async multi downloads

### ISSUE THREAD

**Issue by gaborcsardi** [jeroenooms/curl#76]
I was wondering if it is possible to provide a non-blocking connection to a stream. Ideally, I would occasionally look at the stream if there is some data. Currently I would need to start another (R?) process for this.

Seems like libcurl's multi interface has this functionality. Do you think it is possible to export it to R some way? Polling (or pulling, as libcurl calls it) would be fine for me: when I need data, I'll just see if there is any available.

**Comment by jeroenooms** [comment]
The `base::url` has a `blocking` parameter but I don't think it works currently, that's why I didn't implement it for `curl::curl`. Perhaps they didn't implement it on purpose because most functions assume blocking connections; e.g. if `readLines()` returns no data they assume the connection is exhausted (rather than to try again).

If doing this via the connection API, it's not clear to me how the client would ever know if the connection is exhausted, or it would just keep polling indefinitely. 

It's also not clear what to do if the http request fails all together. You can't really throw an error all of a sudden in the poll function I think...

**Comment by gaborcsardi** [comment]
> The base::url has a blocking parameter but I don't think it works currently, that's why I didn't implement it for curl::curl. Perhaps they didn't implement it on purpose because most functions assume blocking connections; e.g. if readLines() returns no data they assume the connection is exhausted (rather than to try again).

Yes, `readLines()` is messed up. The way I implemented this in `processx` is that there is `is_eof` function that you can call to see if more data *could* arrive. And there is also a `can_read` function just to check if there is data available, without reading it. I am not saying that this is a nice API, but short of events, it is workable.

> If doing this via the connection API, it's not clear to me how the client would ever know if the connection is exhausted, or it would just keep polling indefinitely.

You would need to encapsulate the R connection into another class (or something), and add a method to check for EOF. At least this is one solution.

> It's also not clear what to do if the http request fails all together. You can't really throw an error all of a sudden in the poll function I think...

Why not? But anyway, if the HTTP connection works initially, then it is unlikely that it will suddenly fail, at least this does not happen too often I think. So I don't think this is a critical issue. If the server just closes the connection, then EOF. If the there is some network error, then yes, we can throw an error.

**Comment by jeroenooms** [comment]
> Why not? But anyway, if the HTTP connection works initially, then it is unlikely that it will suddenly fail, at least this does not happen too often I think. So I don't think this is a critical issue. If the server just closes the connection, then EOF. If the there is some network error, then yes, we can throw an error.

But what to do in the initial stage if it is still connecting? You can't let the client wait because it is non-blocking... So the client might `poll` a few times and get no data, and then suddenly get an error because the client couldn't connect to the server at all?

**Comment by gaborcsardi** [comment]
> But what to do in the initial stage if it is still connecting? You can't let the client wait because it is non-blocking... So the client might poll a few times and get no data, and then suddenly get an error because the client couldn't connect to the server at all?

Yes, in general, this is how things should work, I believe. If you open a connection, and read from it successfully, that does not mean that there won't be an error later.

But in practice even if sg is non-blocking, that does not mean that it returns *immediately*. E.g. I am fine if we need to wait for the DNS lookup. :) Maybe even with the HTTP/HTTPS connection, although these both can take time.

Basically, we (you, really :) would just do whatever the libcurl API lets you do. 

Btw. sockets can be non-blocking in base R, and they are used for the `parallel` package for example.

**Comment by gaborcsardi** [comment]
Any prospects on this? I can try to put together a PoC, if you want.

**Comment by jeroenooms** [comment]
How should we signal the user that the connection is completed/exhausted? If we return 0 bytes, it is ambigious if the user should call again or if we're done. How do other connection implementations handle this?

**Comment by gaborcsardi** [comment]
OK, let me start over, now that I have actually looked at (lib)curl a bit.

I would not want to do this for connection (for now), but only for the multi interface. So the API would be similar, almost, but without callbacks. Instead the user is required to poll. Sg. like this:
```r
multi_add(new_handle(url1))
multi_add(new_handle(url2))
...
multi_run(background = TRUE)
while (<condition>) {
  < some R code that I want to run >
  while (multi_any_done()) {
    resps <- multi_get_done()
    <handle resps>
    if (<need-new-download>) multi_add(new_handle(newurl))
  }  
  < some more R code that I want to run >
}
```

So this is actually not very different from your scraping example, but this is a nicer API for me to run code while the downloads are being performed in the background, as opposed to being limited to the callbacks.

A practical example would be a hypothetical :) R package manager that can start installing packages while other packages are being downloaded. But I think that there are a lot potential use cases.

I hope this makes sense.

The next step could be streaming, i.e. getting parts of the downloads, but let's focus on completed requests for now.

**Comment by jeroenooms** [comment]
How do you expect to implement `background = TRUE`? You want to hook this up into the event loop?

**Comment by gaborcsardi** [comment]
No, not at all. I would not touch the event loop if possible. :)

I would just call `curl_multi_perform` in C and that's it. Then it is up to the user to poll using `multi_any_done()`, `multi_get_done()`, or whatever API we come up with, and to eventually destroy the multi. (Although the destroy can also be in the finalizer of course.)

Is there something wrong with this? I am still reading up on libcurl....

**Comment by gaborcsardi** [comment]
Basically I would move your C loop into R. :)

**Comment by jeroenooms** [comment]
Can't you just wrap around `multi_run(timeout = 0)` ?

**Comment by gaborcsardi** [comment]
Maybe, although some changes should be needed, because `curl_multi_perform` is not called at all in this case, AFAICT.

**Comment by gaborcsardi** [comment]
Although it seems that I am wrong.

**Comment by gaborcsardi** [comment]
OK, let me give it a try, this would be a simple and nice solution.

**Comment by gaborcsardi** [comment]
OK, this indeed works, which is great!

It would be still great to have streaming :), but I'll close this one.

```r
done <- 0
cb <- function() {
  print("One is done!")
  done <<- done + 1
}

multi_set(host_con = 2, multiplex = FALSE)
for (i in 1:3){
  multi_add(new_handle(url = httpbin("delay/2")), done = cb, fail = cb)
}

while (TRUE) {
  res <- multi_run(timeout = 0)
  if (res$pending == 0) break;
  print("Here I am, waiting")
  Sys.sleep(0.5)
}

print(done)
```

**Comment by jeroenooms** [comment]
Note in your example above that a call to `multi_run` can result in more than one handle being done.

**Comment by gaborcsardi** [comment]
Sure, thanks for reminding me! 

Btw. how hard would it be to have a callback for chunks of data? If you have that, that's good enough for streaming.

**Comment by jeroenooms** [comment]
I think the `curl()` connection interface is a more natural interface to stream data. It has been specifically designed for this purpose. Hadley has added a simple wrapper `curl_fetch_stream` which uses a callback API.

What would you like to accomplish exactly?

**Comment by jeroenooms** [comment]
I just [added](https://github.com/jeroenooms/curl/commit/10380c276284e6d42c053eca5d19c11fee1ce04c) very basic support for non-blocking connections. So you can do something like:

```r
con <- curl('http://www.httpbin.org/drip?duration=20&numbytes=20&code=200')
open(con, "rb", blocking = FALSE)
repeat{
  buf <- readBin(con, raw(), 100)
  if(length(buf))
    cat("Found some bytes! ", rawToChar(buf))
  Sys.sleep(2)
}
```

However I still don't know how to let the caller know that the connection is done.

**Comment by gaborcsardi** [comment]
Great! 

I think you will need to either
1) have an extra function that checks if the connection is done, or 
2) throw a specific error type (`connectioEof` or sg like this).

IMO 2) is better.

**Comment by jeroenooms** [comment]
Hmz that's a bit ugly. What does base R do? What if the connection closes itself?

**Comment by gaborcsardi** [comment]
But `multi_run(timeout = 0)` does indeed solve most of my use cases, actually:

* checking for outdated packages at the start of an R session, and then set the prompt / or show a desktop notification if they are.
* parallel package downloads and installs.

Couple of similar things....

Originally the R-hub API was streaming the build output, so it would have been good for that, to move the progress indicator instead of blocking, or do sg meaningful

**Comment by gaborcsardi** [comment]
> Hmz that's a bit ugly. What does base R do? What if the connection closes itself?

R does nothing. If you have a non-blocking connection, then you are supposed to implemen EOF externally somehow, as I understand.

There is `isOpen`, but AFAIR that does not work in this case.

I think an EOF error is totally fine, actually. But if you don't like it, then how about returning some special value, i.e. a `character()` with an attribute?

**Comment by gaborcsardi** [comment]
Btw. C sockets "throw" an error if there is no data, but the connection is open. So that's is the other way around. Nevertheless, imo throwing errors to signal a special condition is fine.

**Comment by jeroenooms** [comment]
I cannot control what `readBin` and `readLines` return. The callback only lets you write some data into a buffer and you need to return a `size_t` with how many bytes that you wrote.

**Comment by gaborcsardi** [comment]
Ok, then how about the callback throws an EOF error?

**Comment by gaborcsardi** [comment]
Or EAGAIN / EWOULDBLOCK when the connection is open but there is no data.

**Comment by gaborcsardi** [comment]
Looks like R uses this internally: https://github.com/wch/r-source/blob/af7f52f70101960861e5d995d3a4bec010bc89e6/src/modules/internet/nanohttp.c#L718

**Comment by jeroenooms** [comment]
Hmz that just raises a generic error that is difficult to distinguish from an acutal network error.

I think I found an acceptable solution: https://github.com/jeroenooms/curl/commit/415d7b483c514de85f3350aa8f526a29f4d6e56a. The connection is set to non-readable once it is done. So the caller can check this without actually erroring. If the caller accidentally polls from the exhausted connection it will error anyway.

**Comment by gaborcsardi** [comment]
OK, R has an API actually for signalling EOF on non-blocking connections. It is kinda cumbersome, but workable. 

Just a minute.

**Comment by gaborcsardi** [comment]
So, the API is the `isIncomplete()` function. This returns `FALSE` is the *previous* read is "incomplete", meaning that the connection is not closed. 

In other words, to check if there is more data on a connection, you need to make a `read*`, otherwise `isIncomplete()` does not work, as in itself it does nothing, but looks at the flag that was set by the last read.

After you made a read,  you can call `isIncomplete()`, and it will return `FALSE`, if the connection is closed or there is more data in the input buffer.

From the manual of `?socketConnection`:

```r
     ## Two R processes communicating via non-blocking sockets
     # R process 1
     con1 <- socketConnection(port = 6011, server = TRUE)
     writeLines(LETTERS, con1)
     close(con1)

     # R process 2
     con2 <- socketConnection(Sys.info()["nodename"], port = 6011)
     # as non-blocking, may need to loop for input
     readLines(con2)
     while(isIncomplete(con2)) {
        Sys.sleep(1)
        z <- readLines(con2)
        if(length(z)) print(z)
     }
     close(con2)
```

You need to run this in two R sessions, obviously, because the server blocks.

So maybe you could use the same API for `curl`? 

All you need to do is setting the `incomplete` flag. R does it here for sockets:
https://github.com/wch/r-source/blob/af7f52f70101960861e5d995d3a4bec010bc89e6/src/modules/internet/sockconn.c#L120-L121

**Comment by jeroenooms** [comment]
Sweet, thanks!

**Comment by gaborcsardi** [comment]
So, just to stress this. This is the correct statement, I edited it above as well:

> After you made a read, you can call isIncomplete(), and it will return `FALSE`, if the connection is closed **and** there is **no** more data in the input buffer.

**Comment by jeroenooms** [comment]
This description confuses me though:

> isIncomplete returns a logical value, whether the last read attempt was blocked, or for an output text connection whether there is unflushed output.

It seems like `isIncomplete()` always returns `FALSE` for `base::url()` connections.

**Comment by gaborcsardi** [comment]
The description is very confusing. Forget the unflushed output, that's basically another use for the same flag. The remainder is still confusing. But the socket example is kind of clear, experiment with that, I did that to understand what was going on.

Basically `isIncomplete()` means "can there be more data in this connection"? 

And there are two ways to have more data:
* data in the input buffer, or
* the connection is not closed

> It seems like isIncomplete() always returns FALSE for base::url() connections.

It seems that this flag and this function is only used for non-blocking connections. If you search for incomplete in the R source, you'll see that it is not used for other connections. At least I did not find another use. 

Quite confusingly, it also does not seem to work for non-blocking file reads. I think this is a POSIX thing, files are always blocking. R does some magic to support non-blocking files, e.g. if you keep appending to a file with another process, then you can open it in non-blocking mode in R, and call `readLines()` in a loop on it, but these connection do not seem to use `isIncomplete()`, probably simply because they can't.

Anyway, based on the example in the manual, it should be fine for sockets, and the curl handles are basically sockets, so I think it is natural to do the same.

**Comment by jeroenooms** [comment]
OK I implemented it: https://github.com/jeroenooms/curl/commit/9a670a1bbcd85c9843a9de5b5cd765a9fab97bee. I usually try to mimic `base::url()` but that doesn't seem to work at all for non blocking connections.

**Comment by gaborcsardi** [comment]
Looks good, thanks!

**Comment by gaborcsardi** [comment]
So `req->has_more` is always TRUE if the connection is open? I assume so.

**Comment by jeroenooms** [comment]
That's the value `libcurl` returns for each read to indicate if there is more data to read. So it sounds very similar as the description from `isIncomplete`.

**Comment by gaborcsardi** [comment]
You mean there is more data to read
* from the socket right now?
* or from the socket maybe later?

The latter, right?

Anyway, the httpbin drip is a good test, so I guess it should work fine. :)

**Comment by jeroenooms** [comment]
Yes the latter, it gives the amount of transfers that are still in progress. See [curl_multi_perform](https://curl.haxx.se/libcurl/c/curl_multi_perform.html). In this case it can never be more than 1.

**Comment by gaborcsardi** [comment]
Cool! I am looking forward to using this, and the zero-timeout multi as well! Thanks!

**Comment by jeroenooms** [comment]
This is now on CRAN.

**Comment by gaborcsardi** [comment]
Wow, that was quick! :) Thanks for letting me know.
